### PR TITLE
[russoundserial] Russound Serial Binding initial contribution

### PR DESCRIPTION
**Pull request by holmes**
This binding allows you to control a Russound MCA/CAA system over a
serial connection. There is an existing Russound plugin, but that
works over a network connection on a different protocol (RIO).

This plugin uses a kotlin library to communicate via the serial protocol.
It's available at https://github.com/holmes/russound

Signed-off-by: Jason Holmes <holmes.j@gmail.com> (github: holmes)

### CODE REVIEW

**Comment by holmes** [comment]
I'm not sure what's up w/ the signed-off-by. The line is definitely in my commit!

**Comment by martinvw** [comment]
Maybe the email address does not match the email address you used when committing? I had this problem once too.

**Comment by holmes** [comment]
@martinvw good call. it's using my work author tag instead of my personal one.

**Comment by martinvw** [comment]
Could you please rebase on the current master there was a compilation error in the master 😊

**Comment by holmes** [comment]
rebased nad pushed.

**Comment by holmes** [comment]
@martinvw or @kaikreuzer would y'all take a look and approve?

File: addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/internal/RussoundSerialHandlerFactory.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    return SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
+  }
+
+  @Override
+  protected ThingHandler createHandler(Thing thing) {
+    ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+    if (thingTypeUID.equals(SERIAL_BRIDGE)) {
+      String serialPort = (String) thing.getConfiguration().get("serialPort");
+      File deviceLocation = new File(serialPort);
```
Comment (triller-telekom) [inline addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/internal/RussoundSerialHandlerFactory.java]: This only works on Unix machines, right? On windows we cannot talk to the serial port via a file handle, can we? Or do you handle this somehow in your `SerialCommandSender`? If its a limitation I would briefly mention it in the README file.

File: addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/internal/RussoundSerialHandlerFactory.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        SerialCommandReceiver commandReceiver =
+            new SerialCommandReceiver(translator, new FileInputStream(deviceLocation));
+
+        return new RussoundSerialBridgeHandler((Bridge) thing, commander, commandReceiver,
+            zoneInfoListener);
+      } catch (FileNotFoundException e) {
+        logger.info("Couldn't find serialPort at {}", serialPort, e);
+        return null;
+      } catch (Exception e) {
+        logger.info("Unable to create RussoundSerialBridgeHandler", e);
```
Comment (triller-telekom) [inline addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/internal/RussoundSerialHandlerFactory.java]: A user reading such an error message can probably not do anything about it, so please reduce the severity to `debug`.

File: addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/internal/RussoundSerialHandlerFactory.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import static org.openhab.binding.russoundserial.RussoundSerialBindingConstants.SERIAL_BRIDGE;
+import static org.openhab.binding.russoundserial.RussoundSerialBindingConstants.ZONE_HANDLER;
+
+/**
+ * The {@link RussoundSerialHandlerFactory} is responsible for creating things and thing handlers.
+ *
+ * @author holmes - Initial contribution
+ */
+public class RussoundSerialHandlerFactory extends BaseThingHandlerFactory {
+  private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS =
```
Comment (triller-telekom) [inline addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/internal/RussoundSerialHandlerFactory.java]: Please remove the use of `com.google.common.collect.Sets` and the dependency to this library from the `Manifest`.

Instead you can use:
```
new HashSet<>(Arrays.asList(new String[] { SERIAL_BRIDGE, ZONE_HANDLER }))
```

**Comment by martinvw** [comment]
Or a little shorter:

```java
new HashSet<>(Arrays.asList(SERIAL_BRIDGE, ZONE_HANDLER))
```

File: addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/handler/RussoundSerialZoneHandler.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+  @Override public void handleCommand(ChannelUID channelUID, Command command) {
+    if (command instanceof RefreshType) {
+      scheduleRefresh();
+      return;
+    }
+
+    switch (channelUID.getId()) {
+      case CHANNEL_ZONEPOWER:
+        if (command instanceof OnOffType) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/handler/RussoundSerialZoneHandler.java]: The formatting looks of here, please format using the ESH eclipse profile.

File: addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/handler/RussoundSerialBridgeHandler.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    this.commandReceiver.start();
+  }
+
+  @Override public void dispose() {
+    this.commander.destroy();
+    this.commandReceiver.stop();
+    super.dispose();
+  }
+
+  @Override public void handleCommand(ChannelUID channelUID, Command command) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/handler/RussoundSerialBridgeHandler.java]: Please format is such that @Override has its own line, ie please format all the files you added with the Eclipse ESH profile

File: addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/handler/RussoundSerialZoneHandler.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        logger.info("Unknown/Unsupported Channel id: {}", channelUID.getId());
+    }
+
+    scheduleRefresh();
+  }
+
+  private void scheduleRefresh() {
+    // Every command requires a status request.
+    if (refreshFuture == null || refreshFuture.isCancelled()) {
+      refreshFuture = scheduler.scheduleAtFixedRate(() -> commander().requestStatus(zone), 0, 100,
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/handler/RussoundSerialZoneHandler.java]: Please use [scheduleWithFixedDelay](http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledExecutorService.html#scheduleWithFixedDelay(java.lang.Runnable,%20long,%20long,%20java.util.concurrent.TimeUnit)) which is preferred over scheduleAtFixedRate, see [Coding Guidelines D.3](http://docs.openhab.org/developers/development/guidelines.html#d-runtime-behavior)

Comment (martinvw) [inline addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/handler/RussoundSerialZoneHandler.java]: It looks often, is that really needed?

File: addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/internal/RussoundSerialHandlerFactory.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            Russound.Companion.receiver("russound-serial-bridge", zoneInfoListener);
+        SerialCommandReceiver commandReceiver =
+            new SerialCommandReceiver(translator, new FileInputStream(deviceLocation));
+
+        return new RussoundSerialBridgeHandler((Bridge) thing, commander, commandReceiver,
+            zoneInfoListener);
+      } catch (FileNotFoundException e) {
+        logger.info("Couldn't find serialPort at {}", serialPort, e);
+        return null;
+      } catch (Exception e) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/internal/RussoundSerialHandlerFactory.java]: Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

File: addons/binding/org.openhab.binding.russoundserial/ESH-INF/thing/thing-types.xml [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    <label>Volume</label>
+    <description>Volume level of zone</description>
+  </channel-type>
+  <channel-type id="zoneSource">
+    <item-type>Number</item-type>
+    <label>Source</label>
+    <description>Physical Source Number</description>
+    <state min="0" max="12"/>
+  </channel-type>
+
```
Comment (holmes) [inline addons/binding/org.openhab.binding.russoundserial/ESH-INF/thing/thing-types.xml]: The top ones are `advanced=false`, the bottom are `advanced=true`

File: addons/binding/org.openhab.binding.russoundserial/META-INF/MANIFEST.MF [hunk 9]
```diff
@@ -0,0 +1,26 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: RussoundSerial Binding
+Bundle-SymbolicName: org.openhab.binding.russoundserial;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.2.0.qualifier
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Bundle-ClassPath: ., 
+ lib/api-1.2.0.jar,
```
Comment (holmes) [inline addons/binding/org.openhab.binding.russoundserial/META-INF/MANIFEST.MF]: added

File: addons/binding/org.openhab.binding.russoundserial/META-INF/MANIFEST.MF [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+ lib/kotlin-osgi-bundle-1.1.3.jar
```
Comment (holmes) [inline addons/binding/org.openhab.binding.russoundserial/META-INF/MANIFEST.MF]: added

File: addons/binding/org.openhab.binding.russoundserial/README.md [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+There are 2 supported Thing(s):
+1. Bridge: This is your connection to the controller. You'll specify the serial port once per bridge.
+1. Zones: You'll have a 1:1 mapping of a Zone Thing to a physical zone on your Russound device.
+
+## Discovery and Thing Configuration
+
+There is no auto-discovery. 
+
+#### Bridge
+You must specify the serialPort the Bridge is connected to when setting up the Bridge Thing.
```
Comment (holmes) [inline addons/binding/org.openhab.binding.russoundserial/README.md]: Sure. I left them the parameter names to give a hint as what you should be setting.

File: addons/binding/pom.xml [hunk 12]
```diff
@@ -66,6 +66,7 @@
     <module>org.openhab.binding.rfxcom</module>
     <module>org.openhab.binding.rfxcom.test</module>
     <module>org.openhab.binding.russound</module>
+    <module>org.openhab.binding.russoundserial</module>
```
Comment (holmes) [inline addons/binding/pom.xml]: thanks for the tip, added

File: addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/handler/RussoundSerialBridgeHandler.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
+package org.openhab.binding.russoundserial.handler;
```
Comment (holmes) [inline addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/handler/RussoundSerialBridgeHandler.java]: updated

File: addons/binding/org.openhab.binding.russoundserial/about.html [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+<p>
+    If you did not receive this Content directly from the openHAB community, the Content is
+    being redistributed by another party (&quot;Redistributor&quot;) and different terms and conditions may
+    apply to your use of any object code in the Content. Check the Redistributor's license that was
+    provided with the Content. If no such license exists, contact the Redistributor. Unless otherwise
+    indicated below, the terms and conditions of the EPL still apply to any source code in the Content
+    and such source code may be obtained at <a href="http://www.openhab.org/">openhab.org</a>.
+</p>
+
```
Comment (holmes) [inline addons/binding/org.openhab.binding.russoundserial/about.html]: added

File: addons/binding/org.openhab.binding.russoundserial/README.md [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+  }
+}
+
+```
+
+
+
+## Kotlin
+
+Kotlin is a pretty cool language and the core library that this plugin uses is based on it. Included are OSGI'd versions of the kotlin library for java8.
```
Comment (holmes) [inline addons/binding/org.openhab.binding.russoundserial/README.md]: changed

File: addons/binding/org.openhab.binding.russoundserial/README.md [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+## Discovery and Thing Configuration
+
+There is no auto-discovery. 
+
+#### Bridge
+You must specify the serialPort the Bridge is connected to when setting up the Bridge Thing.
+
+#### Zones
+You must specify the zoneNumber (0-based) this Thing is targeting. This maps directly to the number on the back of the device.
```
Comment (holmes) [inline addons/binding/org.openhab.binding.russoundserial/README.md]: changed

File: addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/handler/RussoundSerialBridgeHandler.java [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+        this.zoneListener = zoneListener;
+    }
+
+    public RussoundCommander getCommander() {
+        return commander;
+    }
+
+    @Override
+    public void initialize() {
+        super.initialize();
```
Comment (holmes) [inline addons/binding/org.openhab.binding.russoundserial/src/main/java/org/openhab/binding/russoundserial/handler/RussoundSerialBridgeHandler.java]: At this point we've established we can make a serial connection, so setting it to `ONLINE` is fine here.

**Comment by holmes** [comment]
@martinvw @kaikreuzer @triller-telekom 

I've addressed all the feedback. Formatting the java files makes it hard to figure out where I changed things that were issues. But I got 'em all.

File: addons/binding/org.openhab.binding.russoundserial/README.md [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+zonePower | Switch | Whether the zone is ON or OFF
+zoneVolume | Dimmer | Volume level of zone 
+zoneSource | Number | The source that's currently playing (0-based)
+zoneBass | Number | Bass Setting (-10 to 10)
+zoneTreble | Number | Treble Setting (-10 to 10)
+zoneBalance | Number | Balance Setting (-10 full left to 10 full right)
+zoneLoudness | Switch | Whether the loudness setting is on or off
+zoneTurnOnVolume | Dimmer | This is a write only setting! Determines the volume when the zone is turned on
+
+## Full Example
```
Comment (triller-telekom) [inline addons/binding/org.openhab.binding.russoundserial/README.md]: You can have a look here:
http://docs.openhab.org/configuration/things.html#defining-things-using-files

or the README files of other bindings also include a full example as well.

**Comment by triller-telekom** [comment]
Oh, one more question that came up while I was browsing through the other bindings... is the current russound binding really that different than your serial one? I mean does it have other things/channels? Otherwise it might make sense to merge the serial related stuff from your binding into the current one.

**Comment by holmes** [comment]
@triller-telekom I intiially tried working this into the existing Russound protocol, but it uses a completely different protocol to communicate. I started to see if I could abstract it out, but it's deeply intertwined in that code.

This protocol/implementation are far simpler so I decided making a new one was a cleaner approach.

**Comment by holmes** [comment]
Last set of changes were to update README w/ .things example and to update the refresh timeout (only called after a command is issued) from 100ms -> 200ms

**Comment by holmes** [comment]
@martinvw @kaikreuzer @triller-telekom

I've addressed all the feedback, squashed commits and am hoping to get this merged! Anything else you'd like me to change?

**Comment by tmrobert8** [comment]
Not to jump in on this but I just noticed it.  The existing Russound binding is already been modified to do all of this (see https://community.openhab.org/t/russound-rnet/27403) by craigh.  The binding uses RNET protocol over a variety of serial connections (native and over several tested IP to serial types).  See [Russound RNET thread](https://community.openhab.org/t/russound-rnet/27403)

I hate seeing work put into this and don't want to really see a duplicate binding happening...

**Comment by holmes** [comment]
Well crap.

The RNET thing is a bit misleading, it's just serial, but this probably looks like it does the same thing.

**Comment by tmrobert8** [comment]
Actually RNET is the protocol and serial is the transport for it (you can actually do RIO protocol over the serial transport also).  Sorry about not seeing this sooner - but feel free to help out with the effort (I haven't heard from craigh for awhile - so you may want to watch that thread and help out!)

**Comment by triller-telekom** [comment]
@holmes: What is the status of this PR? Did you verify that your serial support now exists in the original russound binding? If it is, do you want to close this PR?

I am really sorry that this happened. Its not nice seeing that people put effort into coding something and then there is another one working on the very same thing :(

Maybe in the future we should encourage the developers to submit work in progress (WIP) PRs so people can see what is being worked on at the moment.

**Comment by holmes** [comment]
@triller-telekom I haven't checked the other binding - mine is simple and it works. I couldn't find good documentation nor a build for the other one so I haven't tried.

WIP PRs would be a great thing!

I'll probably just move this to a local repo when the other binding gets merged.

**Comment by LiberatorUSA** [comment]
Based on the conversation above, where can those of us who are in need of the serial RNET support find the correct binding? @wborn @holmes 

People will likely only have 1 Russound unit needing either RIO or RNET. As such having a separate binding for each seems acceptable to me. They are essentially different products just from the same vendor.

**Comment by holmes** [comment]
The binding isn't published anywhere. Looks like it's out of date w/ the latest code anyways.

I'm finishing up a remodel so I don't have my Russound connected right now. Once I get speakers installed I'll take a look at getting this updated.

**Comment by J-N-K** [comment]
@holmes any progress here?

**Comment by sonydevices** [comment]
@holmes 
If you are still interested - you can take over craigs work (I have a copy of it) and finish it off and we could merge...

**Comment by holmes** [comment]
I ended up getting an MCA-88 and run it over ethernet, so I don't even use this binding anymore.

And honestly, i don't have much of a desire to follow up on this. I can maybe find some time to rebase it on master to get it compiling, but I don't have a way to test it.

Gotta strike while the iron is hot. And that was a long time ago.

**Comment by J-N-K** [comment]
@tmrobert8 fondling take over? Otherwise I would close here. 

@holmes Just rebasing doesn‘t help much, sorry. Thanks for your work anyway.

**Comment by sonydevices** [comment]
@J-N-K if I can ever get done with the sony binding, I'll swing back around to this one and implement the serial connection.  Craig had a working version that just needs to be merged with the latest and cleaned up for a merge request.  If anyone wants to volunteer - I can give you the code if you want it done much sooner.

@holmes I though the 88 was locked down now?  Are you a vendor or is that not correct?

**Comment by holmes** [comment]
We can go ahead and close if no one wants to take over.

And yes, they're locked. But I bought a used one that was unlocked.

**Comment by J-N-K** [comment]
Closing as PR is abandoned by original author and noone stepped up to take over.

**Pull request closed without merging**